lib/portage/sync/modules/zipfile/zipfile.py

```diff
@@ -0,0 +1,104 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+# Copyright (C) 2024  Alexey Gladkov <gladkov.alexey@gmail.com>
+
+import os
+import os.path
+import logging
+import zipfile
+import shutil
+import tempfile
+import urllib.request
+
+import portage
+from portage.util import writemsg_level, writemsg_stdout
+from portage.sync.syncbase import SyncBase
+
+
+class ZipFile(SyncBase):
+    """ZipFile sync module"""
+
+    short_desc = "Perform sync operations on GitHub repositories"
+
+    @staticmethod
+    def name():
+        return "ZipFile"
+
+    def __init__(self):
+        SyncBase.__init__(self, "emerge", ">=sys-apps/portage-2.3")
+
+    def sync(self, **kwargs):
+        """Sync the repository"""
+        if kwargs:
+            self._kwargs(kwargs)
+
+        req = urllib.request.Request(url=self.repo.sync_uri)
+
+        info = portage.grabdict(os.path.join(self.repo.location, ".info"))
+        if "etag" in info:
+            req.add_header("If-None-Match", info["etag"][0])
+
+        try:
+            with urllib.request.urlopen(req) as response:
+                with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+                    shutil.copyfileobj(response, tmp_file)
+
+                zip_file = tmp_file.name
+                etag = response.headers.get("etag")
+
+        except urllib.error.HTTPError as resp:
+            if resp.code == 304:
+                writemsg_stdout(">>> The repository has not changed.\n", noiselevel=-1)
+                return (os.EX_OK, False)
+
+            writemsg_level(
+                f"!!! Unable to obtain zip archive: {resp}\n",
+                noiselevel=-1,
+                level=logging.ERROR,
+            )
+            return (1, False)
+
+        if not zipfile.is_zipfile(zip_file):
+            msg = "!!! file is not a zip archive."
+            self.logger(self.xterm_titles, msg)
+            writemsg_level(msg + "\n", noiselevel=-1, level=logging.ERROR)
+
+            os.unlink(zip_file)
+
+            return (1, False)
+
+        # Drop previous tree
+        shutil.rmtree(self.repo.location)
```

Comment:
Like https://github.com/gentoo/portage/pull/991, we could possibly recycle files that have not changed like this:
```diff
--- a/lib/portage/sync/modules/zipfile/zipfile.py
+++ b/lib/portage/sync/modules/zipfile/zipfile.py
@@ -5,2 +5,3 @@ import os
 import os.path
+import filecmp
 import logging
@@ -68,4 +69,10 @@ class ZipFile(SyncBase):
 
-        # Drop previous tree
-        shutil.rmtree(self.repo.location)
+        old = tempfile.mkdtemp(prefix=".old-revision", dir=self.repo.location)
+        oldbasename = os.path.basename(old)
+        for oldfile in os.listdir(self.repo.location):
+            if oldfile != oldbasename:
+                os.rename(
+                    os.path.join(self.repo.location, oldfile),
+                    os.path.join(old, oldfile),
+                )
 
@@ -96,2 +103,15 @@ class ZipFile(SyncBase):
                         shutil.copyfileobj(srcfile, dstfile)
+                        oldfile = os.path.join(
+                            old, os.path.join(old, *parts[strip_comp:])
+                        )
+                        try:
+                            files_equal = filecmp.cmp(dstfile, oldfile, shallow=False)
+                        except Exception:
+                            pass
+                        else:
+                            if files_equal:
+                                os.rename(oldfile, dstfile)
+
+        # Drop previous tree
+        shutil.rmtree(old)
```

Reply:
@zmedico The idea of not deleting the location is good, but the rest will not work.

```python
shutil.copyfileobj(srcfile, dstfile)
```
after this line mtime of `dstfile` will be different and `filecmp.cmp()` will always return `False`.  If we really want to do such a check, then we need to get the checksum of the file.

Reply:
Speaking of `self.repo.location`, you can make the following change:

```diff
diff --git a/lib/portage/sync/modules/zipfile/zipfile.py b/lib/portage/sync/modules/zipfile/zipfile.py
index 3cd210a64..24de2b0c7 100644
--- a/lib/portage/sync/modules/zipfile/zipfile.py
+++ b/lib/portage/sync/modules/zipfile/zipfile.py
@@ -76,7 +76,14 @@ class ZipFile(SyncBase):
             return (1, False)
 
         # Drop previous tree
-        shutil.rmtree(self.repo.location)
+        for name in os.listdir(self.repo.location):
+            if name == ".info":
+                continue
+            f = os.path.join(self.repo.location, name)
+            if os.path.isdir(f):
+                shutil.rmtree(f)
+            else:
+                os.unlink(f)
 
         with zipfile.ZipFile(zip_file) as archive:
             strip_comp = 0
```

Reply:
> +                            files_equal = filecmp.cmp(dstfile, oldfile, shallow=False)

Ops. I missed that `shallow=False` is actually a content comparison.

Reply:
It would be better to do a byte comparison in memory to avoid writing dstfile to disk when possible, but it will get replaced by oldfile so quickly in this case that you would hope the blocks never go flushed to disk.

Reply:
@zmedico I've added an implementation of this, but I implemented it in my own way. The `filecmp.cmp()` always wants to do stat on arguments, which makes the function inapplicable for ZipExtFile. We may independently verify the content is identical.

Reply:
> It would be better to do a byte comparison in memory to avoid writing dstfile to disk when possible, but it will get replaced by oldfile so quickly in this case that you would hope the blocks never get flushed to disk.

@zmedico That's what I did. My current implementation will recreate directories. We don’t have to do this, but then we will need to save the list of files from the zip archive.

Reply:
Thanks this looks great! Let's not bother with directories unless you can come up with a really simple implementation.

Reply:
The implementation will not be difficult, but it will require saving all the paths from the archive in memory. In my case (in the case of local overlay) this is not much, but in the case of the entire portage it will be quite a lot.

I would prefer to leave such changes to follow up PRs and merge it as it is.